pkg/discovery/resolvermgr/grpc_resolver_mgr.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2019 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resolvermgr
+
+import (
+	"math/rand"
+	"strings"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+	"google.golang.org/grpc/naming"
+
+	"github.com/jaegertracing/jaeger/pkg/discovery"
+)
+
+// GRPCResolverManager uses discovery.Notifier to manage grpc servers
+type GRPCResolverManager struct {
+	notifier   discovery.Notifier
+	logger     *zap.Logger
+	discoCh    chan []string // used to receive notifications
+	subsetSize int
+}
+
+type watcher struct {
+	mu         sync.Mutex
+	addrs      []string
+	discoCh    chan []string
+	logger     *zap.Logger
+	subsetSize int
+}
+
+// New creates new GRPCResolverManager.
+func New(
+	notifier discovery.Notifier,
+	logger *zap.Logger,
+	subsetSize int,
+) (*GRPCResolverManager, error) {
+	mgr := &GRPCResolverManager{
+		notifier:   notifier,
+		discoCh:    make(chan []string, 100), // TODO should this number be configurable? What if the number of collectors exceed 100 from discovery service
+		logger:     logger,
+		subsetSize: subsetSize,
+	}
+	// TODO not sure if there's an equivalent way for grpc to maintain connection like what tchannel did?
+
+	notifier.Register(mgr.discoCh)
+	return mgr, nil
+}
+
+// Stop shuts down the manager. It blocks until both bg go-routines exit.
+func (m *GRPCResolverManager) Stop() {
+	if m.discoCh != nil {
+		m.notifier.Unregister(m.discoCh)
+		close(m.discoCh)
+		m.discoCh = nil
+	}
+}
+
+// Resolve implements naming.Resolver interface
+func (m *GRPCResolverManager) Resolve(target string) (naming.Watcher, error) {
+	w := &watcher{
+		discoCh:    m.discoCh,
+		logger:     m.logger,
+		addrs:      []string{},
+		subsetSize: m.subsetSize,
+	}
+	return w, nil
+}
+
+// Next implements naming.Watcher interface
+func (w *watcher) Next() ([]*naming.Update, error) {
+	latestHostPorts := <-w.discoCh
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	w.logger.Info("gRPC naming.Watcher Receives updates", zap.String("hostPorts", strings.Join(latestHostPorts, ",")))
+	var updates []*naming.Update
+
+	// shuffle latestHostPorts
+	r := rand.New(rand.NewSource(time.Now().Unix()))
+	for n := len(latestHostPorts); n > 0; n-- {
+		randIndex := r.Intn(n)
+		latestHostPorts[n-1], latestHostPorts[randIndex] = latestHostPorts[randIndex], latestHostPorts[n-1]
+	}
+	// take the first w.subsetSize
+	if len(latestHostPorts) > w.subsetSize {
+		latestHostPorts = latestHostPorts[:w.subsetSize]
```

Comment:
If you randomly shuffle and then trim the list (e.g. to 3 out of 50), then the subsequent `contains()` check is most likely going to fail, so on every update you'd be reconnecting to new peers, unnecessarily. But if you do `contains()` check before shuffling, then you may create thundering herd if collectors first come up in a small pool. Let's think about how best to address it.

Perhaps Rendezvous Hashing can help. Assign a unique ID to the agent (could be random, as long as it is stable over process lifetime), then sort all peers by `hash(agentID, peerAddress)` and pick the top `N=min-peers`. This way each agent will have its own (unique) preferred ordering of hosts, while the connections will be uniformly distributed across the cluster (and you won't need the `contains()` check at all). It would work even better if instead of host address we'd use our internal instance ID, but I am not sure if we're getting it from Notifier (we could extend the API, in theory).